esmvaltool/diag_scripts/climate_patterns/sub_functions.py

```diff
@@ -0,0 +1,198 @@
+"""Script containing relevant sub-functions for driving scripts.
+
+Author
+------
+Gregory Munday (Met Office, UK)
+"""
+
+import logging
+import multiprocessing as mp
+import os
+from functools import partial
+from pathlib import Path
+
+import iris
+import iris.analysis.cartography
+import iris.coord_categorisation
+import numpy as np
+
+logger = logging.getLogger(Path(__file__).stem)
+
+
+def compute_diagnostic(filename):
+    """Load cube, remove any dimensions of length: 1.
+
+    Parameters
+    ----------
+    filename : path
+        path to load cube file
+
+    Returns
+    -------
+    cube : cube
+        a cube
+    """
+    logger.debug("Loading %s", filename)
+    cube = iris.load_cube(filename)
+    cube = iris.util.squeeze(cube)
+
+    return cube
+
+
+def area_avg(cube, return_cube=None):
```

Comment:
If I replaced this with a preprocessor, wouldn't all my data then be globally averaged? This function is only used to return an average of a portion of the cubes in the recipe, I think most others are left alone in that respect.

Reply:
It is possible to use different preprocessors for different groups of datasets by using the `additional_datasets` to define the datasets that belong to a particular diagnostic/variable instead of using a single list of `datasets` at the top level.

Reply:
The thing is I want both area averaged data and normal gridded data from the _same_ dataset, at different times in the diagnostic - so it wouldn't make much sense to me to duplicate datasets in the recipe file, especially as it's so large already.

Reply:
It looks like the recipe could be made shorter by using just one variable group per variable. Is there a reason why e.g. `tas` is added with three `variable_group`s (`tas_585`, `tas_585_no_tasmax`, and `tas_no_hist_tasmax`) instead of just one?

Reply:
> It looks like the recipe could be made shorter by using just one variable group per variable. Is there a reason why e.g. `tas` is added with three `variable_group`s (`tas_585`, `tas_585_no_tasmax`, and `tas_no_hist_tasmax`) instead of just one?

Essentially, yes. Some models did not have tasmax or tasmin variables available for the 585 scenario on a monthly timestep, so I had to get them from a daily timestep dataset and average monthly. Does that make sense?

Reply:
OK, I think I get that. It is possible to define multiple diagnostics in a recipe and use the [`ancestors` keyword](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/recipe/overview.html#ancestor-tasks) to define which of them feed into a specific diagnostic script, so that might be an option to make this recipe a bit easier to read.

Coming back to the original point: you might get more accurate results from using the `area_statistics` preprocessor function in the recipe, as that will automatically look up the cell area files, which may be more accurate than computing the cell areas (it only does that if it cannot find the cell area file for a specific dataset). If you prefer not to make any changes to the recipe, I would still recommend just importing the [`esmvalcore.preprocessor.area_statistics`](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/api/esmvalcore.preprocessor.html#esmvalcore.preprocessor.area_statistics) function here instead of writing your own.

Reply:
Interesting to know about the ancestor diagnostics.

As for the `area_statistics` preprocessor, I've now updated my diagnostic, replacing my `sf.area_avg` functions with the preprocessor. Thanks!